config/envoyconfig/listeners_main.go

```diff
@@ -0,0 +1,209 @@
+package envoyconfig
+
+import (
+	"context"
+	"time"
+
+	envoy_config_accesslog_v3 "github.com/envoyproxy/go-control-plane/envoy/config/accesslog/v3"
+	envoy_config_core_v3 "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	envoy_config_listener_v3 "github.com/envoyproxy/go-control-plane/envoy/config/listener/v3"
+	envoy_extensions_access_loggers_grpc_v3 "github.com/envoyproxy/go-control-plane/envoy/extensions/access_loggers/grpc/v3"
+	envoy_extensions_filters_network_http_connection_manager "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/network/http_connection_manager/v3"
+	envoy_type_v3 "github.com/envoyproxy/go-control-plane/envoy/type/v3"
+	"google.golang.org/protobuf/types/known/durationpb"
+	"google.golang.org/protobuf/types/known/wrapperspb"
+
+	"github.com/pomerium/pomerium/config"
+)
+
+func (b *Builder) buildMainListener(
+	ctx context.Context,
+	cfg *config.Config,
+	fullyStatic bool,
+) (*envoy_config_listener_v3.Listener, error) {
+	if cfg.Options.InsecureServer {
+		return b.buildMainInsecureListener(ctx, cfg, fullyStatic)
+	}
+	return b.buildMainTLSListener(ctx, cfg, fullyStatic)
+}
+
+func (b *Builder) buildMainInsecureListener(
+	ctx context.Context,
+	cfg *config.Config,
+	fullyStatic bool,
+) (*envoy_config_listener_v3.Listener, error) {
+	li := newListener("http-ingress")
+	li.Address = buildAddress(cfg.Options.Addr, 80)
+
+	// listener filters
+	if cfg.Options.UseProxyProtocol {
+		li.ListenerFilters = append(li.ListenerFilters, ProxyProtocolFilter())
+	}
+
+	filter, err := b.buildMainHTTPConnectionManagerFilter(ctx, cfg, fullyStatic)
+	if err != nil {
+		return nil, err
+	}
+
+	li.FilterChains = []*envoy_config_listener_v3.FilterChain{{
+		Filters: []*envoy_config_listener_v3.Filter{
+			filter,
+		},
+	}}
```

Comment:
Would it be worth extracting this section to a separate method, to share it between buildMainInsecureListener() and buildMainTLSListener()?

(totally optional, please feel free to go either way)